stablehlo/tests/stablehlo_canonicalize_dynamism.mlir

```diff
@@ -1,7 +1,97 @@
+  return %2#0, %2#1 : tensor<1x2xf32>, tensor<3x4xf32>
+}
+
+// -----
+
+func.func @custom_call_failure_attr_number_of_elements(%arg0: tensor<4xf32>) -> tensor<1x2xf32> {
+  // expected-error@+3{{indices_of_shape_operands: number of elements (2) must be equal to the number of operation results (1)}}
+  %0 = stablehlo.constant dense<[1, 2]> : tensor<2xi64>
+  %1 = stablehlo.constant dense<[3, 4]> : tensor<2xi64>
+  %2 = stablehlo.custom_call @foo(%arg0, %0, %1) {
+    indices_of_shape_operands = dense<[1, 2]> : tensor<2xi64>
+  } : (tensor<4xf32>, tensor<2xi64>, tensor<2xi64>) -> tensor<1x2xf32>
+  return %2 : tensor<1x2xf32>
+}
+
+// -----
+
+func.func @custom_call_failure_attr_rank(%arg0: tensor<4xf32>) -> tensor<1x2xf32> {
+  // expected-error@+2{{indices_of_shape_operands: must have rank = 1}}
+  %0 = stablehlo.constant dense<[1, 2]> : tensor<2xi64>
+  %1 = stablehlo.custom_call @foo(%arg0, %0) {
+    indices_of_shape_operands = dense<[[1]]> : tensor<1x1xi64>
+  } : (tensor<4xf32>, tensor<2xi64>) -> tensor<1x2xf32>
+  return %1 : tensor<1x2xf32>
+}
+
+// -----
+
+func.func @custom_call_failure_attr_element_type(%arg0: tensor<4xf32>) -> tensor<1x2xf32> {
+  // expected-error@+2{{indices_of_shape_operands: must have i64 element type}}
+  %0 = stablehlo.constant dense<[1, 2]> : tensor<2xi64>
+  %1 = stablehlo.custom_call @foo(%arg0, %0) {
+    indices_of_shape_operands = dense<[1]> : tensor<1xi32>
+  } : (tensor<4xf32>, tensor<2xi64>) -> tensor<1x2xf32>
+  return %1 : tensor<1x2xf32>
+}
+
+// -----
+
+func.func @custom_call_failure_out_of_bounds_operand_index(%arg0: tensor<4xf32>) -> tensor<1x2xf32> {
+  // expected-error@+2{{indices_of_shape_operands: index #0 (2) must be within bounds for operation operands (from 0 to 2)}}
+  %0 = stablehlo.constant dense<[1, 2]> : tensor<2xi64>
+  %1 = stablehlo.custom_call @foo(%arg0, %0) {
+    indices_of_shape_operands = dense<[2]> : tensor<1xi64>
+  } : (tensor<4xf32>, tensor<2xi64>) -> tensor<1x2xf32>
+  return %1 : tensor<1x2xf32>
+}
+
+// -----
+
+func.func @custom_call_failure_incompatible_result_type(%arg0: tensor<4xf32>) -> tensor<1x2xf32> {
+  // expected-error@+2{{refinement #0 ([1, 1]) must be compatible with operation result ('tensor<1x2xf32>')}}
+  %0 = stablehlo.constant dense<[1, 1]> : tensor<2xi64>
+  %1 = stablehlo.custom_call @foo(%arg0, %0) {
+    indices_of_shape_operands = dense<[1]> : tensor<1xi64>
+  } : (tensor<4xf32>, tensor<2xi64>) -> tensor<1x2xf32>
+  return %1 : tensor<1x2xf32>
+}
+
+// -----
+
+func.func @custom_call_failure_dynamic_shape_operand(%arg0: tensor<4xf32>, %arg1: tensor<2xi64>) -> tensor<1x?xf32> {
+  // CHECK: stablehlo.custom_call @foo(%arg0, %arg1)
+  %0 = stablehlo.custom_call @foo(%arg0, %arg1) {
+    indices_of_shape_operands = dense<[1]> : tensor<1xi64>
+  } : (tensor<4xf32>, tensor<2xi64>) -> tensor<1x?xf32>
+  return %0 : tensor<1x?xf32>
+}
+
+// -----
+
+func.func @custom_call_failure_dynamic_result_type(%arg0: tensor<4xf32>) -> tensor<1x?xf32> {
```

Comment:
Any better name than `_failure_` for these? Or perhaps the other's should be called `_invalid_` and the pass failures are `_failure_`.